cmd/flightctl-worker/main.go

```diff
@@ -85,24 +127,71 @@ func main() {
 		}
 
 		if cfg.Metrics.SystemCollector != nil && cfg.Metrics.SystemCollector.Enabled {
-			if systemMetricsCollector := metrics.NewSystemCollector(ctx, cfg); systemMetricsCollector != nil {
+			if systemMetricsCollector = metrics.NewSystemCollector(ctx, cfg); systemMetricsCollector != nil {
 				collectors = append(collectors, systemMetricsCollector)
+				cleanupFuncs = append(cleanupFuncs, func() error {
+					log.Info("Shutting down system metrics collector")
+					return systemMetricsCollector.Shutdown()
+				})
 			}
 		}
 
 		if len(collectors) > 0 {
 			go func() {
+				log.Info("Starting metrics server")
 				if err := tracing.RunMetricsServer(ctx, log, cfg.Metrics.Address, collectors...); err != nil {
-					log.Errorf("Error running metrics server: %s", err)
+					errCh <- fmt.Errorf("metrics server: %w", err)
+				} else {
+					errCh <- nil
 				}
-				cancel()
 			}()
+			metricsServerStarted = true
 		}
 	}
 
 	server := workerserver.New(cfg, log, store, provider, k8sClient, workerCollector)
-	if err := server.Run(ctx); err != nil {
-		log.Fatalf("Error running server: %s", err)
+
+	go func() {
+		log.Info("Starting worker server")
+		if err := server.Run(ctx); err != nil {
+			errCh <- fmt.Errorf("worker server: %w", err)
+		} else {
+			errCh <- nil
+		}
+	}()
+
+	// Wait for all servers to complete before returning
+	log.Info("Worker service started, waiting for shutdown signal...")
+	serversStarted := 1 // Always start worker server
+	if metricsServerStarted {
+		serversStarted++ // Metrics server was started
+	}
+
+	var firstError error
+	for i := 0; i < serversStarted; i++ {
+		if err := <-errCh; err != nil {
+			if firstError == nil {
+				firstError = err
+				// Cancel context on first error to trigger shutdown of all other servers
+				if !errors.Is(err, context.Canceled) {
+					log.Info("Triggering shutdown of all servers due to error")
+					cancel()
```

Comment:
if the cancel is defered where it is created and when a server errors it returns do we need channels or any of this?